File: lib/math-utils.ts
export function calculateCorrelation(x: number[], y: number[]): number {
  if (x.length !== y.length || x.length === 0) {
    return 0
  }

  const n = x.length
  const meanX = x.reduce((sum, val) => sum + val, 0) / n
  const meanY = y.reduce((sum, val) => sum + val, 0) / n

  let numerator = 0
  let sumXSquared = 0
  let sumYSquared = 0

  for (let i = 0; i < n; i++) {
    const xDiff = x[i] - meanX
    const yDiff = y[i] - meanY

    numerator += xDiff * yDiff
    sumXSquared += xDiff * xDiff
    sumYSquared += yDiff * yDiff
  }

  const denominator = Math.sqrt(sumXSquared * sumYSquared)

  if (denominator === 0) {
    return 0
  }

  return numerator / denominator
}

export function calculateStats(data: number[]) {
  if (data.length === 0) {
    return { average: 0, stdDev: 0 }
  }

  const average = data.reduce((sum, val) => sum + val, 0) / data.length

  const variance =
    data.reduce((sum, val) => {
      const diff = val - average
      return sum + diff * diff
    }, 0) /
    (data.length - 1)

  const stdDev = Math.sqrt(variance)

  return { average, stdDev }
}

export function calculateCovariance(x: number[], y: number[]): number {
  if (x.length !== y.length || x.length === 0) {
    return 0
  }

  const n = x.length
  const meanX = x.reduce((sum, val) => sum + val, 0) / n
  const meanY = y.reduce((sum, val) => sum + val, 0) / n

  const covariance =
    x.reduce((sum, xVal, i) => {
      return sum + (xVal - meanX) * (y[i] - meanY)
    }, 0) /
    (n - 1)

  return covariance
}

File: lib/math_utils.py
def calculate_correlation(x, y):
    """Calculate Pearson correlation coefficient between two arrays"""
    if len(x) != len(y) or len(x) == 0:
        return 0
    
    n = len(x)
    mean_x = sum(x) / n
    mean_y = sum(y) / n
    
    numerator = sum((x[i] - mean_x) * (y[i] - mean_y) for i in range(n))
    sum_x_squared = sum((x[i] - mean_x) ** 2 for i in range(n))
    sum_y_squared = sum((y[i] - mean_y) ** 2 for i in range(n))
    
    denominator = (sum_x_squared * sum_y_squared) ** 0.5
    
    if denominator == 0:
        return 0
    
    return numerator / denominator

def calculate_stats(data):
    """Calculate mean and standard deviation"""
    if not data:
        return {"average": 0, "stdDev": 0}
    
    n = len(data)
    mean = sum(data) / n
    
    if n == 1:
        return {"average": mean, "stdDev": 0}
    
    variance = sum((x - mean) ** 2 for x in data) / (n - 1)
    std_dev = variance ** 0.5
    
    return {"average": mean, "stdDev": std_dev}

def calculate_covariance(x, y):
    """Calculate covariance between two arrays"""
    if len(x) != len(y) or len(x) == 0:
        return 0
    
    n = len(x)
    mean_x = sum(x) / n
    mean_y = sum(y) / n
    
    covariance = sum((x[i] - mean_x) * (y[i] - mean_y) for i in range(n)) / (n - 1)
    return covariance
